### PR TITLE
More advanced table extractor to handle compound queries

### PR DESCRIPTION
**Pull request by crisptrutski**
This makes things much more useful.

Wasn't as bad as I expected!

### CODE REVIEW

File: java/com/metabase/macaw/CompoundTableExtractor.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        final Set<String> cteAliases = new HashSet<>();
+        cteAliasScopes.push(cteAliases);
+
+        if (select.getWithItemsList() != null) {
+            for (WithItem withItem : select.getWithItemsList()) {
+                if (withItem.isRecursive()) {
+                    cteAliases.add(withItem.getAlias().getName());
+                }
+                accTables(withItem.getSelect(), tables, cteAliasScopes);
+                // No hard in adding twice to a set
```
Comment (crisptrutski) [inline java/com/metabase/macaw/CompoundTableExtractor.java:88]:
```suggestion
                // No harm in adding twice to a set
```

File: test/resources/acceptance/compound__nested_cte_sneaky.analysis.edn [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ :source-columns #{{:table "b" :column "x"}
+                   {:table "b" :column "y"}
+                   {:table "b" :column "z"}}
+
+ ;; See https://github.com/metabase/metabase/issues/42586
+ :overrides
+ ;; TODO currently all the sources get cancelled out with the derived columns due to analysis having flat scope.
+ {:ast-walker-1    {:source-columns #{}
+                    :tables         [{:table "a"}]}
+  :compound-select {:tables [{:table "a"}                   ;; not smark enough to shake the tree (this is unused)
```
Comment (crisptrutski) [inline test/resources/acceptance/compound__nested_cte_sneaky.analysis.edn:11]:
```suggestion
  :compound-select {:tables [{:table "a"}                   ;; not smart enough to shake the tree (this is unused)
```

File: java/com/metabase/macaw/CompoundTableExtractor.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        Set<Table> tables = new HashSet<>();
+        Stack<Set<String>> cteAliasScopes = new Stack<>();
+        accTables(select, tables, cteAliasScopes);
+        return tables;
+    }
+
+    private static void accTables(Select select, Set<Table> tables, Stack<Set<String>> cteAliasScopes) {
+        if (select instanceof PlainSelect) {
+            accTables(select.getPlainSelect(), tables, cteAliasScopes);
+        } else if (select instanceof ParenthesedSelect) {
```
Comment (piranha) [inline java/com/metabase/macaw/CompoundTableExtractor.java:45]: is this `ParenthesedSelect` a nested select or what?

Comment (crisptrutski) [reply java/com/metabase/macaw/CompoundTableExtractor.java:45]: Sorry the context has left my brain a bit now 😄 

I think you're right, and from a practical point of view its a SELECT which gets bound to an alias.

**Review by piranha — APPROVED** [review]

**Comment by crisptrutski** [comment]
### Merge activity

* **Oct 25, 3:15 AM EDT**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/metabase/macaw/107).

**Pull request merged**